### ISSUE TITLE
Confusion about process and thread settings on the cluster

### ISSUE THREAD

**Issue by Infinemen** [SmileiPIC/Smilei#50]
Dear Professors，
&emsp;   I am trying to run smilei on a cluster (16 nodes 16 processors per node) with torque (PBS) scheduler. When I run smilei with ./smilei my_namelist.py, everything looks good. And I saw these information on screen: 
``` 
       Number of MPI process: 1;    
       Number of patches : dimension 0 - number_of_patches : 256;      
       Number of thread per MPI process : 32.
```   
&emsp;   In patch_load.txt file:
```
      Total load: 256;     
      patch count: 256.
```
&emsp;   However, when I run smilei on 2 nodes with torque scheduler using the same namelist, the code runs much more slower than the former case even though the cores and memory are double now. And the information list before changes:
```
      Number of MPI process: 2; 
      Number of patches : dimension 0 - number_of_patches : 256;
      Number of thread per MPI process : 16.
``` 
(by adding `export OMP_NUM_THREADS=16` in my .bashrc. I also tried adding `export OMP_NUM_THREADS=32` in .bashrc, which does change the number of thread per MPI process to 32, however, the code still runs slowly. ).

&emsp;   In patch_load.txt file:
```
      Total patches: 128; 
      patches load 128.
```
I changed the number of nodes and ran again using the same namelist, found that the Total load=patch count=256/number of nodes (or number of MPI process, in my submission they equal to each other). What I think is that only the patch count should equal `number of patches / number of nodes`. The Total load should always equals to 256.

&emsp;    What's more, I have talked with the cluster administer. She found that, when run smilei with ./smilei my_namelist.py, the code will run on 16 processors automatically, but when run smilei with torque schedule, the code will run on only 2 processors on the first node leaving the second node unused. It seems like that the number after "mpirun -n" in command `mpirun -n $PBS_NUM_NODES ./smilei my_namelist.py` is the total number of processors required on our cluster, not the total number of process. Finding this, I always set the number equals to the number of processors in submission file and then test smilei with torque schedule. When I use one node, the code runs as fast as running with ./smilei mynamelist.py. But if I use more than one nodes, the MPI will reports communication errors, something like this:
```
At least one pair of MPI processes are unable to reach each other for MPI communications. 
This means that no Open MPI device has indicated that it can be used to communicate between these processes.  
This is an error; Open MPI requires that all MPI processes be able to reach each other. 
This error can sometimes be the result of forgetting to specify the "self" BTL.
Process 1 ([[18428,1],17]) is on host: shenma046
Process 2 ([[18428,1],24]) is on host: shenma044
BTLs attempted: self sm
 
 Your MPI job is now going to abort; sorry.
--------------------------------------------------------------------------
      An error occurred in MPI_Init_thread
--------------------------------------------------------------------------
MPI_INIT has failed because at least one MPI process is unreachable from another.  
This *usually* means that an underlying communication plugin -- such as a BTL or an MTL -- 
has either not loaded or not allowed itself to be used. 
Your MPI job will now abort.

``` 

Waiting for your teaching.

&emsp;   For convenience, here is one of my submission file:
```  
 #PBS -N Smilei
 #PBS -l nodes=2:ppn=16
 #PBS -l walltime=1:00:00
 #PBS -j oe
 #PBS -q parallel11
 module puege
 module load compiler/gcc/7.2.0
 module load mpi/openmpi/1.10.7
 module load hdf5/1.8.16-parallel
 cd $PBS_O_WORKDIR
 mpirun -n $PBS_NUM_NODES ./smilei my_namelist.py
or mpirun -n $ Total_Number_of_Processors  ./smilei my_namelist.py
```

**Comment by beck-llr** [comment]
Hi

First of all you can read the closed issue #36 which is probably the same kind of problem you encounter. Most problems reported to us similar to yours are simply due to a bad openMP threads positioning. If you don't know what I am talking about, refer to your system administrator.  

I strongly encourage not tu use openMP across sockets. Your cluster probably has 2 sockets per node so I would use 2 MPI per nodes instead of one and 8 openMP threads per MPI rank instead of 16.  

There is a typo in your submision file. You probably want to execute `module purge` and not "module puege`.  

The number after -n is the number of MPI proceses you want mpirun to spawn. In your case that should be twice the number of nodes you are using. The number of threads is then set with the environment variables OMP_NUM_THREADS that you have to set to 8 in your case. If you don't set it, the behaviour depends on the system.  

Again, I can not emphasize enough that you have to be very careful about your thread positioning.  Bind them to core, not more than 1 per core and all 8 threads on the same socket.

The error you are reporting at the end of your issue seems to be related to a bad MPI installation. Have you ever run any MPI program on this cluster ?  

Last piece of info: you can completely disregard the "Load" for now. It is just a virtual quantity we use for dynamic load balancing. It is completely independent to your problems. 

Hope this helps solving your problems and see you back here with questions/issues about the code itself.

Cheers

**Comment by Infinemen** [comment]
I know how to set these variables now. The typo is just in the dialog box above. I do want to execute `module purge`.Thanks for correcting this mistake. I have  run MPI program but without OpenMP. So，to make smilei run on our cluster, I have to do some extra settings.
Thank you for your teaching! I will come back soon.
